Dynamic media queries in user-agent shadow DOM stylesheet don't get updated when accessibility settings change

tuankiet65 opened the pull request:
#### ae31aa9f4098d28ea997a4f9e04e585e637be6cb
<pre>
Dynamic media queries in user-agent shadow DOM stylesheet don&apos;t get updated when accessibility settings change
<a href="https://rdar.apple.com/142851714">rdar://142851714</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294905">https://bugs.webkit.org/show_bug.cgi?id=294905</a>

Reviewed by Simon Fraser.

Two events occur when an accessibilty setting changes (like toggling &quot;Smart invert&quot;):

1. Web process got an IPC message that ScreenProperties has changed.
(WebKit::WebProcess::setScreenProperties). This calls
WebKit::WebPage::screenPropertiesDidChange -&gt;
Page::screenPropertiesDidChange -&gt;
Page::setNeedsRecalcStyleInAllFrames -&gt;
Scope::didChangeStyleSheetEnvironment -&gt; calls
Scope::scheduleUpdate on document scope and non-user-agent shadow root scopes.
Scope::scheduleUpdate clears out the resolver so they could be rebuilt with
the new ScreenProperties on the next update. Crucially, resolvers of user-agent
shadow root scopes don&apos;t get cleared. So it&apos;s possible that the resolver of the
document scope is cleared, but not resolvers of user-agent shadow root scopes. (1)

2. Web process got an IPC message that accessibility settings has changed
(Page::accessibilitySettingsDidChange). This calls
Scope::evaluateMediaQueriesForAccessibilitySettingsChange() -&gt;
Scope::evaluateMediaQueries() to re-evaluate dynamic media queries in the document.
To do that, it calls Scope::collectResolverScopes() to collect resolvers for
re-evaluation. The method assumes that if the document scope doesn&apos;t have a
resolver, then the scopes of shadow roots in the document also won&apos;t have a
resolver. This is wrong because of (1)

The effect: When an accessibility setting changes, dynamic media queries in the
document are re-evaluated, except for media queries in user-agent shadow-DOM trees.
This breaks &quot;Smart invert&quot;, since this uses @media (inverted-colors) in
user-agent stylesheet (see Source/WebCore/css/html.css and
Source/WebCore/Modules/modern-media-controls/controls/media-controls.css)

* LayoutTests/fast/media/mq-inverted-colors-ua-style-on-to-off-expected.html: Added.
* LayoutTests/fast/media/mq-inverted-colors-ua-style-on-to-off.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
    - New test is failing on GTK, adding expectation for it.
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::collectResolverScopes):
    - Collect resolvers of shadow root scopes, even when the document scope
      resolver doesn&apos;t exist.
(WebCore::Style::Scope::didChangeStyleSheetEnvironment):
    - Call Scope::scheduleUpdate on all shadow root scopes, including user-agent
      shadow roots.

Canonical link: <a href="https://commits.webkit.org/297386@main">https://commits.webkit.org/297386@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/441c514b3c3f1ede0d6594dfeea3082dd478d407

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111616 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31283 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21757 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117651 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/61887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113578 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31971 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39868 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84801 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/61887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114563 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25519 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100457 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65242 "Found 1 new API test failure: /WPE/TestCookieManager:/webkit/WebKitCookieManager/persistent-storage (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24856 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18592 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61482 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94902 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18661 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120897 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38669 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28726 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93700 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39047 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96715 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93525 "Found 2 new API test failures: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy, /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebView/web-process-crashed (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23842 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38665 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16457 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/34705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38558 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44042 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/38221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41557 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39923 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ea1b23c2a914d6176fdc96eb54745a0788d9b0c4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47119

smfr commented:
Having a layout or API test here would be very valuable for this kind of bug.

tuankiet65 commented:
> Having a layout or API test here would be very valuable for this kind of bug.

The original bug repro is pretty much impossible to replicate in a layout/API test case (I think), but I'll see if there's another way.

anttijk commented on Source/WebCore/style/StyleScope.cpp:
```diff
@@ -791,20 +791,14 @@ auto Scope::collectResolverScopes() -> ResolverScopes
-
     ResolverScopes resolverScopes;
-
-    resolverScopes.add(*resolverIfExists(), Vector<WeakPtr<Scope>> { this });
+    resolverScopes.add(resolver(), Vector<WeakPtr<Scope>> { this });
```
I think the actual issue here is that we bail out if the top level resolver has been cleared but that does not mean there can't be live resolvers in shadow trees. Doing
```
    if (auto* resolver = resolverIfExists())
       resolverScopes.add(*resolver, Vector<WeakPtr<Scope>> { this });
```
seems to fix this.

anttijk replied:
You should make a test case. The original bug says you need dual screens but I can repro just fine with a single screen. See fast/media/mq-inverted-colors-live-update.html, fast/media/mq-inverted-colors-live-update-in-subframes.htm, LayoutTests/accessibility/smart-invert.html for examples.

anttijk replied:
To make a test for this you'll need to figure out how to delete the document style resolver without deleting the shadow tree ones. I think you can achieve this by using any non-dynamic media query (one without MediaQueryDynamicDependency in MediaQueryFeature.cpp) on a document stylesheet.

tuankiet65 replied:
I've changed it to your suggestion and added a test case.

anttijk commented:
> Scope::collectResolverScopes should recreate resolver if on doesn't exist
> [rdar://142851714](https://rdar.apple.com/142851714)
> https://bugs.webkit.org/show_bug.cgi?id=294905

It would be better if the bug title described the problem rather than a solution.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c7c4770d26991ca95dbddd7f672a29286c6c0742. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47119

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/181d5d6842c70798d4b17101f14df055e23c2c29. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47119

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7c873619668b417edbd051d23ad2edf6d7bfe8d1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47119

smfr commented on LayoutTests/fast/media/mq-inverted-colors-ua-style-on-to-off.html:
```diff
@@ -0,0 +1,70 @@
+
+        window.internals.settings.forcedColorsAreInvertedAccessibilityValue = "on";
+
+        testRunner.runUIScript(uiScript, () => {
```
Can you just use `await UIHelper.renderingUpdate()` instead? Include `[...]/resources/ui-helper.js` to get that.

tuankiet65 replied:
I tried and it doesn't work.

The style resolution code doesn't re-evaluate media queries unless it has received an "accessibility settings changed" notification. A user changing the settings would trigger that, but to do that in the testcase requires a function in UIScript to simulate it. I suppose `UIHelper.renderingUpdate` just waits for the next frame but doesn't send the notification, so nothing would be invalidated/updated.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/58c76a863c8b1f7d72d835a0cff86ffcf3c7d6fd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47119

tuankiet65 committed:
Dynamic media queries in user-agent shadow DOM stylesheet don't get updated when accessibility settings change
rdar://142851714
https://bugs.webkit.org/show_bug.cgi?id=294905

Reviewed by Simon Fraser.

Two events occur when an accessibilty setting changes (like toggling "Smart invert"):

1. Web process got an IPC message that ScreenProperties has changed.
(WebKit::WebProcess::setScreenProperties). This calls
WebKit::WebPage::screenPropertiesDidChange ->
Page::screenPropertiesDidChange ->
Page::setNeedsRecalcStyleInAllFrames ->
Scope::didChangeStyleSheetEnvironment -> calls
Scope::scheduleUpdate on document scope and non-user-agent shadow root scopes.
Scope::scheduleUpdate clears out the resolver so they could be rebuilt with
the new ScreenProperties on the next update. Crucially, resolvers of user-agent
shadow root scopes don't get cleared. So it's possible that the resolver of the
document scope is cleared, but not resolvers of user-agent shadow root scopes. (1)

2. Web process got an IPC message that accessibility settings has changed
(Page::accessibilitySettingsDidChange). This calls
Scope::evaluateMediaQueriesForAccessibilitySettingsChange() ->
Scope::evaluateMediaQueries() to re-evaluate dynamic media queries in the document.
To do that, it calls Scope::collectResolverScopes() to collect resolvers for
re-evaluation. The method assumes that if the document scope doesn't have a
resolver, then the scopes of shadow roots in the document also won't have a
resolver. This is wrong because of (1)

The effect: When an accessibility setting changes, dynamic media queries in the
document are re-evaluated, except for media queries in user-agent shadow-DOM trees.
This breaks "Smart invert", since this uses @media (inverted-colors) in
user-agent stylesheet (see Source/WebCore/css/html.css and
Source/WebCore/Modules/modern-media-controls/controls/media-controls.css)

* LayoutTests/fast/media/mq-inverted-colors-ua-style-on-to-off-expected.html: Added.
* LayoutTests/fast/media/mq-inverted-colors-ua-style-on-to-off.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
    - New test is failing on GTK, adding expectation for it.
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::collectResolverScopes):
    - Collect resolvers of shadow root scopes, even when the document scope
      resolver doesn't exist.
(WebCore::Style::Scope::didChangeStyleSheetEnvironment):
    - Call Scope::scheduleUpdate on all shadow root scopes, including user-agent
      shadow roots.

Canonical link: https://commits.webkit.org/297386@main
